File: zig-backend/src/model/Application.zig
const std = @import("std");

pub const applicationTableName = "applications";

const Date = std.time.Timestamp;

id: i32,
applying_date: Date,
employer: []const u8,
webpage: []const u8,
position: []const u8,
contact_person: []const u8,
contact_person_gender: []const u8,
acknowledgement_date: Date,
interview_date: Date,
declination_date: Date,
acknowledged_occured: bool,
interview_occured: bool,
declination_occured: bool,
contact_type_id: i32,
status_id: i32,

pub fn init(data: anytype) @This() {
    return .{
        .id = data.id,
        .applying_date = try getDate(data.applying_date),
        .employer = data.employer,
        .webpage = data.webpage,
        .position = data.position,
        .contact_person = data.contact_person,
        .contact_person_gender = data.contact_person_gender,
        .acknowledgement_date = try getDate(data.acknowledgement_date),
        .interview_date = try getDate(data.interview_date),
        .declination_date = try getDate(data.declination_date),
        .acknowledged_occured = data.acknowledged_occured,
        .interview_occured = data.interview_occured,
        .declination_occured = data.declination_occured,
        .contact_type_id = data.contact_type_id,
        .status_id = data.status_id,
    };
}

fn getDate(input: anytype) !Date {
    if (input) |date| {
        return date;
    } else |str| {
        return try std.time.parseISODateTime(str);
    }
}

File: model/Application.ts
import * as v from 'valibot'

const dateSchema = v.pipe(v.union([v.date(), v.string()]), v.transform((value) => new Date(value)))
const boolSchema = v.pipe(v.union([v.boolean(), v.number()]), v.transform((value) => value === 1))

export const applicationTableName = 'applications'
export const applicationSchema = v.object({
    id: v.optional(v.number()),
    applying_date: dateSchema,
    employer: v.string(),
    webpage: v.string(),
    position: v.string(),
    contact_person: v.optional(v.string(), '-'),
    contact_person_gender: v.optional(v.string(), 'm'),
    acknowledgement_date: v.optional(dateSchema),
    interview_date: v.optional(dateSchema),
    declination_date: v.optional(dateSchema),
    acknowledged_occured: v.optional(boolSchema, false),
    interview_occured: v.optional(boolSchema, false),
    declination_occured: v.optional(boolSchema, false),
    contact_type_id: v.number(),
    status_id: v.number()
})

export type ApplicationInput = v.InferInput<typeof applicationSchema>
export type Application = v.InferOutput<typeof applicationSchema>
/*
export type DateField = 'applying_date' | 'acknowledgement_date' | 'interview_date' | 'declination_date'
export type BooleanField = 'acknowledged_occured' | 'interview_occured' | 'declination_occured'
export type NumberField = 'contact_type_id' | 'status_id'
export type StringField = 'employer' | 'webpage' | 'position' | 'contact_person' | 'contact_person_gender'

export class Application {
    id!: number
    applying_date!: Date
    employer!: string
    webpage!: string
    position!: string
    contact_person!: string
    contact_person_gender!: string
    acknowledgement_date!: Date | null
    interview_date!: Date | null
    declination_date!: Date | null
    acknowledged_occured!: boolean
    interview_occured!: boolean
    declination_occured!: boolean
    contact_type_id!: number
    status_id!: number

    constructor(data: any | null | undefined) {
        // Parse with Valibot schema (which now includes defaults)
        const result = v.safeParse(applicationSchema, data)
        if (!result.success) {
            throw new Error(`Invalid application data: ${JSON.stringify(result.issues, null, 2)}`)
        }
        // Assign validated data
        Object.assign(this, result.output)
    }
}
*/